File: example/example-java/src/main/java/org/cibseven/community/bpm/data/example/process/OrderApprovalInstanceFactory.java
package org.cibseven.community.bpm.data.example.process;

import org.cibseven.bpm.engine.RuntimeService;
import org.cibseven.bpm.engine.runtime.ProcessInstance;
import org.cibseven.bpm.engine.variable.VariableMap;
import org.springframework.stereotype.Component;

import java.util.UUID;

import static org.cibseven.bpm.engine.variable.Variables.createVariables;
import static org.cibseven.community.bpm.data.example.process.OrderApproval.ORDER_ID;

/**
 * Factory to create instance factory.
 */
@Component
public class OrderApprovalInstanceFactory {

  /**
   * Runtime service to access CIB seven API.
   */
  private final RuntimeService runtimeService;

  /**
   * Constructs the factory.
   *
   * @param runtimeService runtime service to use.
   */
  public OrderApprovalInstanceFactory(RuntimeService runtimeService) {
    this.runtimeService = runtimeService;
  }

  /**
   * Start new approval process.
   *
   * @param orderId id of an order.
   * @return instance supplier.
   */
  public OrderApprovalInstance start(String orderId) {
    VariableMap vars = createVariables();
    ORDER_ID.on(vars).set(orderId);
    ProcessInstance instance = runtimeService.startProcessInstanceByKey(OrderApproval.KEY, "order-" + UUID.randomUUID().toString(), vars);
    return new OrderApprovalInstance(instance);
  }

}

File: example/example-kotlin/src/main/kotlin/process/OrderApprovalInstanceFactory.kt
package org.cibseven.community.bpm.data.example.kotlin.process

import org.cibseven.community.bpm.data.builder.VariableMapBuilder
import org.cibseven.community.bpm.data.example.kotlin.process.OrderApproval.Variables.ORDER_ID
import org.cibseven.bpm.engine.RuntimeService
import org.springframework.stereotype.Component
import java.util.*

/**
 * Instance factory.
 */
@Component
class OrderApprovalInstanceFactory(
  private val runtimeService: RuntimeService
) {

  /**
   * Starts the approval process.
   */
  fun start(id: String): OrderApprovalInstance {
    val vars = VariableMapBuilder().set(ORDER_ID, id).build()
    val instance = runtimeService.startProcessInstanceByKey(OrderApproval.KEY, "order-${UUID.randomUUID()}", vars)
    return OrderApprovalInstance(instance)
  }

}
